src/stan/mcmc/hmc/nuts_classic/base_nuts_classic.hpp

```diff
@@ -0,0 +1,249 @@
+#ifndef STAN_MCMC_HMC_NUTS_BASE_NUTS_HPP
+#define STAN_MCMC_HMC_NUTS_BASE_NUTS_HPP
+
+#include <stan/interface_callbacks/writer/base_writer.hpp>
+#include <boost/math/special_functions/fpclassify.hpp>
+#include <stan/mcmc/hmc/base_hmc.hpp>
+#include <stan/mcmc/hmc/hamiltonians/ps_point.hpp>
+#include <algorithm>
+#include <cmath>
+#include <limits>
+#include <string>
+#include <vector>
+
+namespace stan {
+  namespace mcmc {
+
+    struct nuts_util {
+      // Constants through each recursion
+      double log_u;
+      double H0;
+      int sign;
+
+      // Aggregators through each recursion
+      int n_tree;
+      double sum_prob;
+      bool criterion;
+    };
+
+    // The No-U-Turn Sampler (NUTS) with the
+    // original slice sampler implementation
+    template <class Model, template<class, class> class Hamiltonian,
+              template<class> class Integrator, class BaseRNG>
+    class base_nuts_classic:
+      public base_hmc<Model, Hamiltonian, Integrator, BaseRNG> {
+    public:
+      base_nuts_classic(Model &model, BaseRNG& rng):
+        base_hmc<Model, Hamiltonian, Integrator, BaseRNG>(model, rng),
+        depth_(0), max_depth_(5), max_delta_(1000),
+        n_leapfrog_(0), divergent_(0), energy_(0) {
+      }
+
+      ~base_nuts_classic() {}
+
+      void set_max_depth(int d) {
+        if (d > 0)
```

Comment:
I'd prefer an exception if `d <= 0`, as it indicates a programmer error on the outside.  If it's only called internally, you don't need the check as you can make sure it's called properly.